File: 0501-1000/0839. Similar String Groups.java
// Hard
// DFS -> Apply for the data size, Optimal: Union Find
// Time:O(n² * m), Space:O(n)
// https://leetcode.cn/problems/similar-string-groups/

class Solution {
    public int numSimilarGroups(String[] strs) {
        int n = strs.length;
        boolean[] vis = new boolean[n];
        int group = 0;

        for (int i = 0; i < n; i++) {
            if (!vis[i]) {
                group++;
                dfs(strs, vis, i);
            }
        }
        return group;
    }

    private void dfs(String[] strs, boolean[] vis, int index) {
        vis[index] = true;
        for (int i = 0; i < strs.length; i++) {
            if (!vis[i] && isSimilar(strs[index], strs[i])) {
                dfs(strs, vis, i);
            }
        }
    }

    private boolean isSimilar(String a, String b) {
        int diff = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                diff++;
                if (diff > 2)
                    return false;
            }
        }
        return diff == 2 || diff == 0;
    }
}

/**
 * 两个字符串相似表示可以通过交换两个字符得到相同的字符串 => 他们属于同一组 => 问有多少这样的组合
 * 可以把两个字符串看作是图中的节点，如果他们相似，则连成一条边，然后用DFS搜索连通数量
 */

File: 0501-1000/0839. Similar String Groups.py
# Time:O(n² × L), Space:O(n²) 

from typing import List

class Solution:
    def numSimilarGroups(self, strs: List[str]) -> int:
        n = len(strs)
        graph = [[] for _ in range(n)]
        vis = [False] * n

        def is_similar(a, b):
            if a == b:
                return True
            diff = []
            for i in range(len(a)):
                if a[i] != b[i]:
                    diff.append(i)
                    # they can't swap
                    if len(diff) > 2:
                        return 
            return len(diff) == 2 and a[diff[0]] == b[diff[1]] and a[diff[1]] == b[diff[0]]

        for i in range(n):
            for j in range(i + 1, n):
                if is_similar(strs[i], strs[j]):
                    graph[i].append(j)
                    graph[j].append(i)
        
        def dfs(u):
            vis[u] = True
            for v in graph[u]:
                if not vis[v]:
                    dfs(v)

        cnt = 0
        for i in range(n):
            if not vis[i]:
                dfs(i)
                cnt += 1
        
        return cnt 
    